lib/db.c

```diff
@@ -74,12 +74,79 @@ static const uint32_t GRN_TABLE_PAT_KEY_CACHE_SIZE = 1 << 15;
       grn_string_get_normalized(ctx, nstr, &key, &key_size, NULL);\
       block\
       grn_obj_close(ctx, nstr);\
+    } else {\
+      GRN_DEFINE_NAME(table);\
+      ERR(GRN_INVALID_ARGUMENT,\
+          "[key][normalize] failed to normalize: <%.*s> <%.*s>",\
+          (int)name_size, (const char *)name,\
+          (int)key_size, (const char *)key); \
     }\
   } else {\
     block\
   }\
 } while (0)
 
+#define INSPECT_TABLE_LIMITED_WITH_KEY(message, table, key, key_size) do { \
+  grn_obj buffer;\
+  GRN_TEXT_INIT(&buffer, 0);\
+  grn_inspect_limited(ctx, &buffer, table);\
+  if (key && key_size > 0) {\
+    grn_id domain_id = table->header.domain;\
+    grn_obj *domain;\
+    domain = grn_ctx_at(ctx, domain_id);\
+    if (grn_obj_is_text_family_type(ctx, domain)) { \
+      ERR(GRN_UNKNOWN_ERROR,    \
```

Comment:
There are extra spaces between ',' and `\`.

Reply:
We should use suitable return code as much as possible.

Reply:
I've fixed to use proper GRN_XXX.